
## Reinforcement Learning
__Reinfrocemnt Learning__ is a type of machine learning tasks where decisions makers are trying to maximize long-term rewards or minimize long-term costs. In the leanring proces, the decision-makers learn how to maximize rewards or minimize costs throught the interaction expereinces in the enviroment (key feature  is guided by reward signals or cost signals).

_example_ robot in a maze observes its surroundigns and treis to move, stupid moves waste time and enegy while smart moves lead it out of the maze. in this example ove is the action after its obersvation and time and enrgy sit spends is the cost (negative rewards) 




### key terms

<ls>

 1. __Reward__ (or __cost__): the goal of RL is to maximize the long-term reward, or minimize the long-term cost. in the example of the maze, the time and energy wasted  gereante negative rewards. The robot wants to get postiive rewards while avoiding ngeative rewards.

 2. __Policy__ : Decision-maker will act according to their obsercations, and the way of making decisions is called _policy_.  It' is the policy that RL algorithm learns. RL algorithm changes policy to maximize the total rewards.

</ls>



### Differences between RL and other ML 

<ls>

1. __Supervised learning__: Supervised learning, learners know the correct action, and can learn by comapring each decision with the correct decisiojn. in RL the __the correct action is unknown__ so the algorithm can only learn from reward signals.  RL wants to maximize the total reward for a whole period and usually seeks long-term superiority. 

    
    * supervised learning comapres the action for each deicison so it is myopic
     
    * supervised learning usually hopes the learning result can be used in unknown data and the results should generalize well

    * the polcy obtained by RL can bused in the training enviroment

    * supervised learning can be used for predition or classifcations such as predicting stock prices  or identiify the contents of images, but RL doe snot excel in such tasks

2. __Unsupervised Learning__:  Unsupervised learning aims to discover the structure that the data implies. RL has explict quantitative goals .i.e. the rewards. unsupervised learning can be used in cluster, RL does not excel in clustering

</ls>



### Agent-Enviroment Interface
**Agent-enviroment interface** is a tool to analyze an RL system
1. __Agent__ is the decision-maker and learning. As __decision-maker__ the agent can decide what action to take. As __learner__ the agent can adjust its policty. There can be one agent or many agents in an RL system
2. __Enviroment__ inclues everytthign in RL system that agents do not include. it is what the aents interacts with
  
_example_ At work i can decide what to do, and leanrr to adjust polcy that determins what to do. Those belong to the *agent*. Meanwhile my health, tiredness and starvatyion are things i can not control directly. so they belon gto the *enviroment*. The agent can respond to the enviroment. Say i may go to bed if i feel tired,

_note_ while  RL systmes are allyzed using agent-enviroment interface, in most cases it is not imperativie to use agent-enviroment-interfaces.
* Agent observe teh enviroment and get the __observation__ denoted *O*
* Agent use some polcy to determin the __action__ denoted *A*
* after being impacted by the action, the enviroment signals _rewards__ dennoted _R_ and cahange its __state__ denoted *S*

_note_ States, observations and actions are not necessarily nmeric values, they can be stuff like feeling hungry and eating (denoted in sans-serif). Rewards are always numerical values (scalers denoted in serif)

_note_ most RL systems are casual which means what happened eariler can only impact what happens aftewards. Sequential decision makign  can be indexed on time denote _t_, the state at time _t_ can be denoted as $S_{t}$. Independent events do not require a time elemetn, such as rolling a dice and the reward is the outcome of the dice. 

when the intereaction timing between an agent and enviroment is countable it is called a discrete-time agent-enviroment interface. the intearction timing index can be mapped to $t = 1,2,3,...$

_note_ some liteature denote teh reard signal after teh action $A_t$ as $R_t$ while others might denote it as $R_{t+1}$ (<- will used $t+1$ moving on)



### Taxonomy of RL
There are various ways to classify RL based on tasks, or based on algorithms


#### Task-based


- single/multi-agent tasks (__MALR__): in a 
__single-agent tasks__ there is only one agent who can observe, in multi-agent, each agent may just osberse its own obsercation. in multi-agent  taks agents may or may not communicate. Each agent may have differnt goals that might conflict.
- episodic/continous tasks : an __episodic tasks__ has start and end. For example a game of Go  is an episode tasks where the board is empty at the beginning and full at the end. a __continous tasks__ has neither beginning nor end.  for example: te scheduler of the public cloud servies needs to process without interruption forever
- discrete/continious-time tasks : __discrete-time tasks__ : time index is countably infinite or the time index is finite wher e$T$ can be a random variable. __continious-time__ if the time interaction between agent and the enviroment is uncountalbe i.e. all non-negative real numbers or non-empty coutnious subset. __note__ classifcation is not excluseive, additonally the intervals between two interactions can be stochastic and the inteval can futher impact sequetional evolutions.
- discrete/continious action space: __discrete action space__ - coutnable outcomes (only moving north,east, south, west). __continous actions space__ - uncountable outcome (moving between $[ 0\degree,  360\degree ]$)
- deterministic/ stochastic enviroment: __deterministic enviroment__ - the enviroment does not change, for example a robot in a fixed maze. __stochastic enviroment__ - the enviroment is stochastic. for example the robot is in a randomally generated maze
- fully/partially observalbe enviroment: __fully observable enviroment__ - agent can see the whole enviroment. for example in the game of Go because players can see the whole board. __partially obsevable__ agents can onlys ee part of the enviroment. for example in poker, players do not know what cards the opponets have


#### Algorithem-based

- on/off policy learning: __on-policy__ - learns from the policy that is ued to play (example Imagine you're trying to teach a robot to navigate a maze. In on-policy learning, the robot learns based on the actions it is currently taking. It's like learning to cook by trying out different recipes yourself.) __off-policy__ - can lean from records of games with old polcies ( learning from someone else's experience. In this approach, the robot might watch another robot navigate the maze and learn from its actions. It doesn't have to follow the same policy as the robot it's observing.). in offline tasks, agents can not apply the latest action  so only __off-policy__ may be used in off line tasks
- model- based/free learning : __Model-based__ reinforcement learning uses a model of the environment to simulate outcomes and plan actions, while __model-free__ reinforcement learning learns directly from interactions with the environment without building a model. Model-based methods can be more efficient but are often more complex to implement than model-free methods. example model-based game of Go the rules of the gam  are the model and we know them prior.
- Monte Carlo/Temporal difference Update: __Monte Carlo update__ suitable for episode tasks, __monte carlo update__ waits until the end of the eispode to update the value of the state, and learn from the estimate of eapisodes returns.  __temporal update__ does not need to finish the eipside, it can se the intermediate information to learn. 
- value-based, policy-based or actor-critic: __Valued-based__:  use state values and action values to denote the return we can get on the condition on a certian state and/or action.  __policy based__ change the policy directly wihtout the helo of balues. __actor_critic__ methods combine __value-based__ and __polciy_based__ methods. __policy gradent__ are the most important type of __policy-based__ algorithms
- deep/ classical reinforcment learning:  if an algorithm use neral networks as a componet, it leverages __deep learning (DL)__ and is called __deep RL__. example video game AI can use convlutional neural networks to process video game images and solve RL taks, since it uses both RL and neural networks it is a __Deep Reinforcment learning algorithm__


some sytems need to consider more than one taks. Learning fro multiple tasks is called __multi-task reinfrocment learning (MTLR)__ if mayn taks differy onlyi n parameters of the reward function such __MTLR__ my be called __goal-condtiional RL__. example in a multi-task system each tasks may consist of each robot going to a task-specific destinaition. the destination can be called the goal of the tasts and those tasks are goal conditional taks. 

Another type of leanring is a technique where an agent is developed for a particular task is reused as the starting point for a model on a second task. This is particularly useful when the second task has limited data called __transfer reinforcemnt learning__. 

if we want the resulting agent to easily adapt to new taks we are conducing __meta reinforcment learning__

if the taks is evolving and our agent  needs to adapt to the changin gtasks thi sis called __online reinforcment learning__ or __lifelong reinformcnet learning__.

_NOTE_ __online RL__ is not the antonym of __offline RL__. __Offline RL__ is learning wihtout inteactions with enviroments , and only leanrs from historical records of inteactions usually in batches (__batch reinforcment learning__)

addtionaly __passive RL/ polcy evaluation__ is a kidn of taks that evaluates the preformance of a fixed polyc, while __active RL / policy optimization__ tries to implve the polcy. 


### Performance Metrics

* __Long-tern Reward__: this indicates the long-term reward that an algorithm can achieve when it is applied to the task. Many statistics are derived from long-term rewards such as mean and st.dev. We usually want larger returns and episode rewards. We also may want small Std.devs
* __optimality and convergence__: we perfer the algorithms that can converage to the optimal
* __sample complexity__ :  it is usually costly and time-consuming to interact with enviroments we we refer to learn from as few samples as possible. Somealgorithms can make use of the enviroment models or historical records to learn so they have smaller sample compleixty. 

>_note_ Some model-based algorithms can even solve the tasts wihtout learning from samples. strictly speaking,s uch algorithms are not RL algorithms. Algorithms for offline tasks learn from historical records, they do not interact with the enviroments. Therefore the "online" sample complxity of these offline RL is 0, however offline RL tasks usually get a fixed number of samples so the offline sample complexity is critical.

* __regeret__: regert is defined as the summation of episode regerts during the training process. __Regret__ is define das the difference between teh acatual episode reweard and the optimal episode reward after the training completes. we prefer small regert values. Some RL algorithms learn from scratct and accumlate regert , others are pretrained in one cheaper enviroment then trained in a costly enviroment. We only care about the regert in the costly enviroment. example: an agent can be trained in a simulation enviroment before it is formally trianed in the real workd becuase failure in the real world is costly. 

* __scalability__: some RL algorithms can use multipleworkers to accelerate learning.Since RL tasks are usually time-consuming an dcomputation-intensive, it is an advauntage to ahve scalability to sepped up learning.

* __convergence speed, training time and complexity__: Theorist care whether an RL algorithm converages and how fast it converges. Time complexity can also be used to indicate the speed of convergence. Convergence is a good property and we prefer small time complexity

> _note_ Convergence in reinforcement learning (RL) refers to the process where an RL algorithm's performance stabilizes, meaning that the agent's learned policy consistently yields optimal or near-optimal rewards over time. This typically occurs when the agent has sufficiently explored the environment and refined its strategy based on the feedback received

* __occupied space and space complexity__ : we perfer small space occupation, some use cases particualy care about the space for storing historical records while other use cases care more about the storeage on GRAM

### Case-study Agent _enviroment Interface in GYM

https://gymnasium.farama.org/

> _note_ Gym has been unmaintained since 2022, and amongst other critical missing functionality does not support Numpy 2.0, and the documentation website has been taken offline. Gymnasium is the maintained drop in replacement for Gym from the original Gym team if you're on the latest version of Gym.

https://gymnasium.farama.org/introduction/migration_guide/
|compponent|old|new|impact|
|----------|---|---|------|
|Package Import | `import gym` | `import gymnasium as gym` | all code|
|Environment Reset | `obs = env.reset()` | `obs, info = env.reset()` | training loops|
|Random Seeding | `env.seed(42)` | `env.reset(seed=42)` | Reproducibility|
|Step Function | `obs, reward, done, info = env.step(action)` | `obs, reward, terminated, truncated, info = env.step(action)` | RL algorithms|
|Episode Ending | `while not done:` | `while not (terminated or truncated):` | Training loops|
|Render Mode | `env.render(mode="human")` | `gym.make(env_id, render_mode="human")` | Visualization|
|Time Limit Detection | `info.get('TimeLimit.truncated')` | `truncated`  _return value_ | RL algorithms|
|Value Bootstrapping | `target = reward + (1-done) * gamma * next_value` | `target = reward + (1-terminated) * gamma * next_value` | RL correctness|


* Toy Text: this category includes some simple task that can be represented by strings
* Classic Control : this category contains some simple kenetic tasks
* Box2D; the tasks in this cateogry uses libeary Box2D to create and visulize objects
* Atarai: this category includes atarai games


### Using Gym

In [1]:
import gymnasium as gym
# note pip install gymnaisum first https://github.com/openai/gym

we can use `gym.make()` to get an enviroment object. Every task in gym has an ID string in the format 'XxxxYyyy-vN' such as "CartPole-v0' and "Taxi_v3". _note_ each version of tasks has differnt behavoires

In [4]:
env = gym.make('CartPole-v0')

c:\Users\endro\miniconda3\Lib\site-packages\gymnasium\envs\registration.py:512: DeprecationWarning: WARN: The environment CartPole-v0 is out of date. You should consider upgrading to version `v1`.
  logger.deprecation(


In [10]:
[[key,value] for key,value in gym.envs.registry.items()][0:3]

[['CartPole-v0',
  EnvSpec(id='CartPole-v0', entry_point='gymnasium.envs.classic_control.cartpole:CartPoleEnv', reward_threshold=195.0, nondeterministic=False, max_episode_steps=200, order_enforce=True, disable_env_checker=False, kwargs={}, namespace=None, name='CartPole', version=0, additional_wrappers=(), vector_entry_point='gymnasium.envs.classic_control.cartpole:CartPoleVectorEnv')],
 ['CartPole-v1',
  EnvSpec(id='CartPole-v1', entry_point='gymnasium.envs.classic_control.cartpole:CartPoleEnv', reward_threshold=475.0, nondeterministic=False, max_episode_steps=500, order_enforce=True, disable_env_checker=False, kwargs={}, namespace=None, name='CartPole', version=1, additional_wrappers=(), vector_entry_point='gymnasium.envs.classic_control.cartpole:CartPoleVectorEnv')],
 ['MountainCar-v0',
  EnvSpec(id='MountainCar-v0', entry_point='gymnasium.envs.classic_control.mountain_car:MountainCarEnv', reward_threshold=-110.0, nondeterministic=False, max_episode_steps=200, order_enforce=True, d

a task has its observation space and its action space. For an enviroment `env`, `env_observation_space` records its observation space while `env.action_space` records it's action space. The gym class `gym_spaces.Box` can be used to represent a space of `np.array's`. 

for a space with finite elements we can also use the class `gym.spaces.Discrete` to represent a space of $\mathbb{Z}$  

_example_: the observation space of `CartPole-v0` is `Box(4,0)` which is an `np.array` of the shape of 4 and the action space `cartPole-v0` is `Discrete(2)`  so teh action can be an int value within {0,1}. A box object uses members $low$ and $high$ to record the range of values. A `Discrete` ojbect uses the member $n$ to record the number of elements in the shape. 

To interact with the enviroment objt we need to initalize the enviroment with:



In [13]:
obs, info = env.reset()
obs

array([ 0.03607941,  0.01681996,  0.02956897, -0.02935553], dtype=float32)

`obs` returns the intial observation , `info`is a dictioanry.

For each timestamp, we can use the member `env.step()` to interact with the enviroment object

>`step()` - Updates an environment with actions returning the next agent observation, the reward for taking that actions, if the environment has terminated or truncated due to the latest action and information from the environment about the step, i.e. metrics, debug info.
>> Changed in version 0.26: The Step API was changed removing done in favor of terminated and truncated to make it clearer to users when the environment had terminated or truncated which is critical for reinforcement learning bootstrapping algorithms.

the `step()` method has one argument `action`: a `ActType` provided by the agent to update the enviroment state

the `step()` method returns the following 5 values:
* `observation` : a `ObsType` which is a value sharing the same meaning with the first return value of `env.reset()`. in other words an element of the environment’s observation_space as the next observation due to the agent actions. An example is a numpy array containing the positions and velocities of the pole in CartPole.
* `reward`:  a `float` value
* `terminated`:  a `bool` value indicating whether the epsidoe has been ended. teh enviroments in Gym are mostly episodic. When `terminated` is `True` it means that the epsidoe has finsihed. In this case we can call `obs,info = env.reset()` to start the next episode
* `truncated`:  a `bool` value indicating whether the epsidoe has been trucnated. We can limit the number of steps in an episode in both episodic tasks and sequential tasks. So that sequential tasks become episodic tasks with max number of steps. when The epsidoe reaches its max steps the `trunctaed` indicator becomes `True`useful when resource limitations such as memory is nout enough or the data exceeds valid range 
* `info`: a `Dict` – Contains auxiliary diagnostic information (helpful for debugging, learning, and logging). shares the same meaning with the second return value of `env.reset()`

Each time we call `step.env()` we move the time line one step. In order to interact with the enviroment sequentially, we usually put it in a loop

after `env.reset()` or `env.step()` we may use the following code to visulize the enviroment


In [15]:
env.render()

c:\Users\endro\miniconda3\Lib\site-packages\gymnasium\envs\classic_control\cartpole.py:250: UserWarning: WARN: You are calling render method without specifying any render mode. You can specify the render_mode at initialization, e.g. gym.make("CartPole-v0", render_mode="rgb_array")
  gym.logger.warn(


after using the enviromen, we can use the following code to close the enviroment

In [16]:
env.close()

_note_ `env.render()` may render the enviroment in a new window. In this case, the best way to close the window is to call `env.close()` directly rather than clicking the 'X' since this might cause crashes.

When researchers use the enviroments in Gym as benchmark enviroments, they usually run 100 episdoes successively and calcualte the average episdoe rewards. They chose to run 100 epsiodes rather than other number of episodes such as 128  just as common practice. Some enviroments also have refrence threshold for average episode rewards. An algorithm is said to solve the task if an donly if the average episode rewards exceeds teh refence threshold.

if a tasks has a refrence episode reward threshold, the threshold is stored in

```env.spec.reward_threshold```

### Example: MountainCar Finite action space

This example is he classical task called 'Mountain Car' where under-powered car must drive up a steep hill. Since gravity is stronger than the car's engine, even at full throttle, the car cannot simply accelerate up the steep slope. The car is situated in a valley and must learn to leverage potential energy by driving up the opposite hill before the car is able to make it to the goal at the top of the rightmost hill. In Gym there are two versions of this `MountainCar-v0` and `MountainCarContinious-v0`.



In [4]:
import gymnasium as gym
import logging
import numpy as np

#note logger would not output information, i had to add this to make it output
logging.getLogger().setLevel(logging.INFO)

env = gym.make('MountainCar-v0')
for key in vars(env.spec):
    logging.info('%s: %s' , key, vars(env.spec)[key])
for key in vars(env.unwrapped):
    logging.info('%s: %s' , key, vars(env.unwrapped)[key])    


INFO:root:id: MountainCar-v0
INFO:root:entry_point: gymnasium.envs.classic_control.mountain_car:MountainCarEnv
INFO:root:reward_threshold: -110.0
INFO:root:nondeterministic: False
INFO:root:max_episode_steps: 200
INFO:root:order_enforce: True
INFO:root:disable_env_checker: False
INFO:root:kwargs: {}
INFO:root:additional_wrappers: ()
INFO:root:vector_entry_point: None
INFO:root:namespace: None
INFO:root:name: MountainCar
INFO:root:version: 0
INFO:root:min_position: -1.2
INFO:root:max_position: 0.6
INFO:root:max_speed: 0.07
INFO:root:goal_position: 0.5
INFO:root:goal_velocity: 0
INFO:root:force: 0.001
INFO:root:gravity: 0.0025
INFO:root:low: [-1.2  -0.07]
INFO:root:high: [0.6  0.07]
INFO:root:render_mode: None
INFO:root:screen_width: 600
INFO:root:screen_height: 400
INFO:root:screen: None
INFO:root:clock: None
INFO:root:isopen: True
INFO:root:action_space: Discrete(3)
INFO:root:observation_space: Box([-1.2  -0.07], [0.6  0.07], (2,), float32)
INFO:root:spec: EnvSpec(id='MountainCar-v0', 

the results show that:
* the action space is `action_space: Discrete(3)` so the action will be an `int` value within the set ${0,1,2}$
* teh observation space is `observation_space: Box([...], [...], (2,), float32)` so the observation will be a `np.array` with shape(2,)
* the max steps in an episode is `max_episode_steps: 200`
* refrence episode reward value `reward_threshold: -110.0`. An algorithm is said to solve the task if it's average episode reward over 100 successive epsisodes excceds this threshold

next we prepare the agent tfor interacting with the enviroment. Genrally gym does not include agents so we need to implement an agent by ourselves. The class `ClosedFormAgent` below will solve this tasks. it is a simplifed agent that can only decide but not learn. Therefore, not an RL but just to illustrate the point

In [2]:
class ClosedFormAgent:
    def __init__(self, _):
        pass

    def reset(self, mode=None):
        pass

    def step(self, observation, reward, terminated):
        position, velocity = observation
        lb = min(-0.09 * (position + 0.25) ** 2 + 0.03,
                0.3 * (position + 0.9) ** 4 - 0.008)
        ub = -0.07 * (position + 0.38) ** 2 + 0.07
        if lb < velocity < ub:
            action = 2 # push right
        else:
            action = 0 # push left
        return action

    def close(self):
        pass


agent = ClosedFormAgent(env)

once we have an agent and enviorment interact with each other, the finaly plart is to make the function `play_episode` with the following arguments

* `env`: the enviroment - object
* `agent`: the agent - object
* `seed`: seed of RNG for epsidoe - `None` or `int`
* `mode`: this agent can not learn so thjis is no use, set it to `None` - `None` or `'train'`
* `render`: indicating wether to visulize the interaction or not - `bool`

the function returns `episode-reward` and `elapse_steps`. `episode_reward` is a `float` indcating the episode reward. `elapsesd_ste` is an `int` inddicating the number of steps in an epside

In [6]:
def play_episode(env, agent, seed=None, mode=None, render=False):
    observation, _ = env.reset(seed=seed)
    reward, terminated, truncated = 0., False, False
    agent.reset(mode=mode)
    episode_reward, elapsed_steps = 0., 0
    while True:
        action = agent.step(observation, reward, terminated)
        if render:
            env.render()
        if terminated or truncated:
            break
        observation, reward, terminated, truncated, _ = env.step(action)
        episode_reward += reward
        elapsed_steps += 1
    agent.close()
    return episode_reward, elapsed_steps




in order to evaluate the performance of the agent systematically calculate the average epsidoe rewards over 100 successive epsides. 

In [7]:
logging.info('==== test ====')
episode_rewards = []
for episode in range(100):
    episode_reward, elapsed_steps = play_episode(env, agent)
    episode_rewards.append(episode_reward)
    logging.info('test episode %d: reward = %.2f, steps = %d',
            episode, episode_reward, elapsed_steps)
logging.info('average episode reward = %.2f ± %.2f',
        np.mean(episode_rewards), np.std(episode_rewards))

INFO:root:==== test ====
INFO:root:test episode 0: reward = -104.00, steps = 104
INFO:root:test episode 1: reward = -105.00, steps = 105
INFO:root:test episode 2: reward = -97.00, steps = 97
INFO:root:test episode 3: reward = -105.00, steps = 105
INFO:root:test episode 4: reward = -96.00, steps = 96
INFO:root:test episode 5: reward = -106.00, steps = 106
INFO:root:test episode 6: reward = -105.00, steps = 105
INFO:root:test episode 7: reward = -88.00, steps = 88
INFO:root:test episode 8: reward = -84.00, steps = 84
INFO:root:test episode 9: reward = -128.00, steps = 128
INFO:root:test episode 10: reward = -105.00, steps = 105
INFO:root:test episode 11: reward = -130.00, steps = 130
INFO:root:test episode 12: reward = -104.00, steps = 104
INFO:root:test episode 13: reward = -147.00, steps = 147
INFO:root:test episode 14: reward = -105.00, steps = 105
INFO:root:test episode 15: reward = -104.00, steps = 104
INFO:root:test episode 16: reward = -104.00, steps = 104
INFO:root:test episode 1

we see the agent `ClosedFormAgent` can attain an average approximately -105, which excceds the reward threshold of -110. we can concolude this agent solves the tasks